omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_solr_validator.rb

```diff
@@ -48,6 +66,172 @@ def verify_sane_reindex_sleep_times
     end
   end
 
+  def verify_one_search_index
+    if elasticsearch_enabled? && solr_enabled?
+      fail_with <<~EOM
+
+      The #{CHEF_SERVER_NAME} is configured to enable both
+      Elasticsearch and Solr. This is an unsupported configuration.
+
+      Please update #{CHEF_SERVER_CONFIG_FILE} such that only one
+      search index is enabled.
+      EOM
+    end
+  end
+
+  # If an external search provider (ES or Solr) was explicitly enabled
+  # by the user, then we expect that both internal search services
+  # should be disabled.
+  def verify_es_disabled_if_user_set_external_solr
+    # NOTE(ssd) 2020-05-13: This might impact the people who we gave
+    # pre-release access to.
+    if cs_solr_attr['external'] && elasticsearch_enabled?
+      fail_with <<~EOM
+
+      The #{CHEF_SERVER_NAME} is configured to use an external search
+      index but the internal Elasticsearch is still enabled. This is
+      an unsupported configuration.
+
+      To disable the internal Elasticsearch, add the following to #{CHEF_SERVER_CONFIG_FILE}:
+
+          elasticsearch['enable'] = false
+
+      To use the internal Elasticsearch, consider removing the configuration
+      entry for opscode_solr4['external'].
+
+      EOM
+    end
+
+    if solr_enabled? && external?
+      fail_with <<~EOM
+
+      The #{CHEF_SERVER_NAME} is configured to use an external search
+      index but the internal Solr is still enabled. This is an
+      unsupported configuration.
+
+      To disable the internal Solr, add the following to #{CHEF_SERVER_CONFIG_FILE}:
+
+          opscode_solr4['enable'] = false
+
+      Alternatively, if you are attempting to use the internal Solr,
+      add the following to #{CHEF_SERVER_CONFIG_FILE}:
+
+          opscode_solr4['enable'] = true
+          opscode_solr4['external'] = false
+
+      EOM
+    end
+  end
+
+  # The user might have a combination of solr, rabbitmq or
+  # opscode-expander explicitly enabled in their config.
+  #
+  # If we are using the new internal elasticsearch, this is an
+  # unsupported configuration.
+  #
+  # To test: setup standalone cs with 13.2.0 add
+  # opscode_solr4['enable'] = true
+  # rabbitmq['enable'] = true
+  # opscode_expander['enable'] = true to chef-server.rb
+  # then upgrade to 13.3
+  def verify_unused_services_are_disabled_if_using_internal_es
+    if elasticsearch_enabled?
+      if cs_solr_attr['enable']
+        fail_with <<~EOM
```

Comment:
Instead of immediately calling fail_with, we could track the failures in some array or string and then fail once at the bottom to avoid too much iteration.

Reply:
Done